internal/core/src/cachinglayer/SegmentLoadGuard.cpp

```diff
@@ -0,0 +1,600 @@
+        bool eviction_state_changed = (config_.eviction_enabled != config.eviction_enabled);
+        config_ = config;
+        
+        if (eviction_state_changed) {
+            if (config_.eviction_enabled) {
+                StartEvictionThread();
+            } else {
+                StopEvictionThread();
+            }
+        }
+    }
+    LOG_INFO("SegmentLoadGuard config updated: cache_ratio={}, eviction_enabled={}", 
+             config_.cache_ratio, config_.eviction_enabled);
+}
+
+LoadGuardConfig SegmentLoadGuard::GetConfig() const {
+    std::shared_lock<std::shared_mutex> lock(config_mutex_);
+    return config_;
+}
+
+void SegmentLoadGuard::SetPhysicalLimits(int64_t memory_limit, int64_t disk_limit) {
+    memory_limit_.store(memory_limit);
+    disk_limit_.store(disk_limit);
+    LOG_INFO("Physical limits updated: memory={}GB, disk={}GB", 
+             memory_limit / (1024.0 * 1024.0 * 1024.0),
+             disk_limit / (1024.0 * 1024.0 * 1024.0));
+}
+
+std::pair<int64_t, int64_t> SegmentLoadGuard::GetPhysicalLimits() const {
+    return {memory_limit_.load(), disk_limit_.load()};
+}
+
+bool SegmentLoadGuard::CanLoadSegment(const Segment& segment) const {
+    auto config = GetConfig();
+    
+    // Calculate current shallow usage with cache ratio applied
+    ResourceUsage current_shallow_usage;
+    current_shallow_usage.memory_bytes = shallow_inevictable_.load().memory_bytes + 
+                                        static_cast<int64_t>(shallow_evictable_.load().memory_bytes * config.cache_ratio);
+    current_shallow_usage.file_bytes = shallow_inevictable_.load().file_bytes + 
+                                      static_cast<int64_t>(shallow_evictable_.load().file_bytes * config.cache_ratio);
+    
+    // Calculate segment's shallow usage
+    ResourceUsage segment_shallow_usage;
+    segment_shallow_usage.memory_bytes = segment.GetInevictableSize() + 
+                                       static_cast<int64_t>(segment.GetEvictableSize() * config.cache_ratio);
+    segment_shallow_usage.file_bytes = 0; // Disk usage calculated separately if needed
+    
+    auto total_shallow = current_shallow_usage + segment_shallow_usage;
+    
+    // Check against logical limits
+    auto [memory_limit, disk_limit] = GetPhysicalLimits();
+    int64_t logical_memory_limit = static_cast<int64_t>(memory_limit * config.overload_percentage);
+    int64_t logical_disk_limit = static_cast<int64_t>(disk_limit * config.overload_percentage);
+    
+    if (total_shallow.memory_bytes > logical_memory_limit || 
+        total_shallow.file_bytes > logical_disk_limit) {
+        return false;
+    }
+    
+    // Check physical limits as additional safety
+    auto physical_usage = GetCurrentPhysicalUsage();
+    return CheckPhysicalLimits(segment_shallow_usage);
+}
+
+bool SegmentLoadGuard::LoadSegment(std::shared_ptr<Segment> segment) {
+    if (!CanLoadSegment(*segment)) {
```

Comment:
shouldn't this entire `CanLoadSegment, if so UpdateShallowUsage` be guarded by a lock?